### PR TITLE
Improve const/ref this decorators for list methods

### LINKED ISSUE #14715: Unable to update elements of a list within a map
### Summary of Problem

Updates to the elements of a list within a map seem to be dropped. See the example code below.

**Source Code:**

```chapel
use Map;
use List;

var m: map(string, list(int));

m['a'] = new list(int);

writeln(m);
// {a: []}

m['a'].append(1);
writeln(m);
// {a: [1]}

m['a'][1] = 2;
writeln(m);
// {a: [1]}
// but I expected:
// {a: [2]}
```

Taking a peek at the `Map.chpl` source code, I noticed there are 2 `this` methods:

```chpl
proc this(k: keyType) ref { .. }
proc const this(k: keyType) const { .. }
```

It turns out the `const this` is getting called in my program above, which explains why the updates are not happening. Assuming this is actually a bug and not a feature, I suspect the issue lies in compiler resolution.

Related to https://github.com/chapel-lang/chapel/issues/14474

### Configuration Information

- Output of `chpl --version`: chpl version 1.21.0 pre-release (4f4e86253a)

### PR DESCRIPTION
**Pull request by mppf**
Resolves #14715.

Note that until #12944 is fixed, the iterators won't work correctly with return-intent-overloading.

Reviewed by @dlongnecke-cray - thanks!

- [x] full local testing

### CODE REVIEW

**Comment by mppf** [comment]
@daviditen - fyi

**Comment by mppf** [comment]
There are some test failures to attend to
```
[Error compilation failed for library/packages/UnitTest/Launcher_Test (compopts: 1)]
[Error matching compiler output for mason/chplVersion/mason-chpl-version-build]
[Error compilation failed for mason/chplVersion/mason-chpl-version-update (compopts: 1)]
[Error matching compiler output for mason/mason-example-with-opts/mason-example]
[Error matching compiler output for mason/mason-example/mason-example]
[Error matching compiler output for mason/mason-test/mason-test]
[Error compilation failed for mason/mason-update-toml-error/mason-chpl-version-update (compopts: 1)]
[Error compilation failed for mason/pkgconfig-tests/pcTest (compopts: 1)]
[Error matching compiler output for mason/run/mason-run]
[Error matching compiler output for mason/subdir-commands/mason-run]
[Error compilation failed for mason/masonUpdateTest (compopts: 1)]
[Error matching compiler output for studies/shootout/reverse-complement/bharshbarg/revcomp-buf]
[Error matching compiler output for studies/shootout/reverse-complement/bharshbarg/revcomp-line]
```

File: modules/standard/List.chpl [hunk 1]
```diff
@@ -361,6 +361,12 @@ module List {
       ref result = array[itemIdx];
       return result;
     }
+    inline proc ref _getRef(idx: int) ref {
```
Comment (dlongnecke-cray) [inline modules/standard/List.chpl]: In general, for cases like this where we have to "forcibly" tell the compiler that a method is ref, is it always enough to say that "if this method returns a ref, then the receiver is ref as well"? 

In my mind `ref` is used when the receiver is mutated inside the method, but I guess that's not entirely correct.

Comment (dlongnecke-cray) [reply modules/standard/List.chpl]: I guess `this` indexing routines are a special case, though...

**Comment by benharsh** [comment]
FYI, this PR appears to have caused the following reverse-complement regression:
https://chapel-lang.org/perf/chapcs/?startdate=2019/12/16&enddate=2020/02/13&graphs=reversecomplementshootoutbenchmark

@dlongnecke-cray : can you take a look at this when you have a chance (not urgent)?

Pinging @chapel-lang/perf-team as well

**Comment by bradcray** [comment]
I was curious what happened here, and if it's not obvious from the previous comment (it wasn't to me), this PR changed the code for the two revcomp versions that regressed (so it's not necessarily surprising that their performance changed).  Specifically, some lists are now passed around by `ref` in the benchmark (via task and function intents) when they were passed around by default intent (which I assume was `const ref`?) previously.

This made me curious whether the explicit task intents could be made into `const ref`, but it seems that they can't because we modify the lists' elements in these lines in each benchmark:

```chapel
    ref d1 = data[start], d2 = data[end];
    (d1, d2) = (table[d2], table[d1]);
```

So it seems likely that we were getting away with something before that we shouldn't have been able to (assuming that `const ref` on a list should prohibit modifying a list's elements in addition to prohibiting adding/removing elements from the list as a whole, which seems correct to me).

**Pull request merged**